File: ops/digest/digest.py
# Daily digest: taxonomy from watcher SQLite + outcome JSONL. No invented numbers.

import argparse
import json
import sqlite3
import sys
from collections import Counter
from pathlib import Path


def main() -> int:
    p = argparse.ArgumentParser()
    p.add_argument("--sqlite", required=True)
    p.add_argument("--jsonl", required=True)
    args = p.parse_args()
    sqlite = Path(args.sqlite)
    jsonl = Path(args.jsonl)
    if not sqlite.is_file():
        print(f"FAIL sqlite missing: {sqlite}", file=sys.stderr)
        return 1
    if not jsonl.is_file():
        print(f"FAIL jsonl missing: {jsonl}", file=sys.stderr)
        return 1
    con = sqlite3.connect(f"file:{sqlite.as_posix()}?mode=ro", uri=True)
    try:
        n = con.execute("SELECT COUNT(*) FROM liquidations").fetchone()[0]
    except sqlite3.Error as e:
        print(f"FAIL sqlite: {e}", file=sys.stderr)
        return 1
    names: Counter[str] = Counter()
    jsonl_rows = 0
    halts = 0
    with jsonl.open(encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            jsonl_rows += 1
            try:
                obj = json.loads(line)
            except json.JSONDecodeError as e:
                print(f"FAIL jsonl: {e}", file=sys.stderr)
                return 1
            if "halt" in obj:
                halts += 1
            name = obj.get("outcome") or obj.get("name")
            if name:
                names[str(name)] += 1
    print(f"sqlite_rows={n}")
    print(f"jsonl_rows={jsonl_rows}")
    print(f"NotTracked={names.get('NotTracked', 0)}")
    print(f"HealthWrong={names.get('HealthWrong', 0)}")
    print(f"Declined={names.get('Declined', 0)}")
    print(f"halts={halts}")
    print("by_outcome=" + json.dumps(dict(names), sort_keys=True))
    if names.get("NotTracked", 0) or names.get("HealthWrong", 0):
        print("ALARM_COUNTS_NONZERO", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/liq-obs/src/digest.rs
//! Taxonomy counts from shadow JSONL + watcher SQLite. Missing files fail; no guessed counts.

use std::collections::BTreeMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

use rusqlite::Connection;
use serde::Deserialize;

use crate::error::{ObsError, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DigestReport {
    pub sqlite_rows: u64,
    pub jsonl_rows: u64,
    pub by_outcome: BTreeMap<String, u64>,
    pub not_tracked: u64,
    pub health_wrong: u64,
    pub declined: u64,
    pub halt_lines: u64,
}

#[derive(Deserialize)]
struct OutcomeLine {
    #[serde(default)]
    outcome: Option<String>,
    #[serde(default)]
    name: Option<String>,
}

fn inc(map: &mut BTreeMap<String, u64>, k: &str) -> Result<()> {
    let e = map.entry(k.to_owned()).or_insert(0);
    *e = e.checked_add(1).ok_or_else(|| {
        tracing::error!("digest counter overflow");
        ObsError::Json("counter overflow".into())
    })?;
    Ok(())
}

pub fn digest_paths(sqlite: &Path, jsonl: &Path) -> Result<DigestReport> {
    if !sqlite.is_file() {
        tracing::error!(path = %sqlite.display(), "digest sqlite missing");
        return Err(ObsError::Io(format!("missing {}", sqlite.display())));
    }
    if !jsonl.is_file() {
        tracing::error!(path = %jsonl.display(), "digest jsonl missing");
        return Err(ObsError::Io(format!("missing {}", jsonl.display())));
    }
    let db = Connection::open(sqlite)?;
    let sqlite_rows: i64 = db.query_row("SELECT COUNT(*) FROM liquidations", [], |r| r.get(0))?;
    let sqlite_rows = u64::try_from(sqlite_rows).map_err(|_| {
        tracing::error!("sqlite COUNT negative");
        ObsError::Sqlite("negative count".into())
    })?;

    let file = File::open(jsonl)?;
    let reader = BufReader::new(file);
    let mut by_outcome = BTreeMap::new();
    let mut jsonl_rows = 0u64;
    let mut not_tracked = 0u64;
    let mut health_wrong = 0u64;
    let mut declined = 0u64;
    let mut halt_lines = 0u64;
    for line in reader.lines() {
        let line = line?;
        if line.is_empty() {
            continue;
        }
        jsonl_rows = jsonl_rows.checked_add(1).ok_or_else(|| {
            tracing::error!("jsonl row counter overflow");
            ObsError::Json("overflow".into())
        })?;
        let v: serde_json::Value = serde_json::from_str(&line)?;
        if v.get("halt").is_some() {
            halt_lines = halt_lines
                .checked_add(1)
                .ok_or(ObsError::Json("overflow".into()))?;
        }
        let name = v
            .get("outcome")
            .and_then(|x| x.as_str())
            .or_else(|| v.get("name").and_then(|x| x.as_str()));
        if let Some(n) = name {
            inc(&mut by_outcome, n)?;
            match n {
                "NotTracked" => {
                    not_tracked = not_tracked
                        .checked_add(1)
                        .ok_or(ObsError::Json("overflow".into()))?;
                }
                "HealthWrong" => {
                    health_wrong = health_wrong
                        .checked_add(1)
                        .ok_or(ObsError::Json("overflow".into()))?;
                }
                "Declined" => {
                    declined = declined
                        .checked_add(1)
                        .ok_or(ObsError::Json("overflow".into()))?;
                }
                _ => {}
            }
        } else if let Ok(ol) = serde_json::from_value::<OutcomeLine>(v.clone()) {
            if let Some(n) = ol.outcome.or(ol.name) {
                inc(&mut by_outcome, &n)?;
            }
        }
    }
    Ok(DigestReport {
        sqlite_rows,
        jsonl_rows,
        by_outcome,
        not_tracked,
        health_wrong,
        declined,
        halt_lines,
    })
}

impl DigestReport {
    #[must_use]
    pub fn render(&self) -> String {
        format!(
            "sqlite_rows={}\njsonl_rows={}\nNotTracked={}\nHealthWrong={}\nDeclined={}\nhalts={}\nby_outcome={:?}\n",
            self.sqlite_rows,
            self.jsonl_rows,
            self.not_tracked,
            self.health_wrong,
            self.declined,
            self.halt_lines,
            self.by_outcome
        )
    }
}
